src/mainboard/sunxi/nezha/main/src/main.rs

```diff
@@ -59,45 +58,48 @@ const DTB_ADDR: usize = MEM + DTB_OFFSET;
 const EI: usize = 12;
 type MyLzss = lzss::Lzss<EI, 4, 0x00, { 1 << EI }, { 2 << EI }>;
 
-fn decompress() {
-    // check for Device Tree header, d00dfeed
-    let dtb = unsafe { read_volatile(DTB_ADDR as *const u32) };
-    if dtb != 0xedfe0dd0 {
-        panic!("DTB looks wrong: {:08x}\n", dtb);
-    } else {
-        print!("DTB looks fine, yay!\n");
-    }
-
+fn decompress(compressed_addr: usize, target_addr: usize, payload_size: usize) {
     // first four bytes are the compressed size
-    let in_ptr = (LINUXBOOT_TMP_ADDR + 4) as *const u8;
-    let out_ptr = LINUXBOOT_ADDR as *mut u8;
-    let compressed_size = unsafe { read_volatile(LINUXBOOT_TMP_ADDR as *const u32) };
+    let in_ptr = (compressed_addr + 4) as *const u8;
+    let out_ptr = target_addr as *mut u8;
+    let compressed_size = unsafe { read_volatile(compressed_addr as *const u32) };
     print!(
-        "Decompress {} bytes from {:?} to {:?}\n",
-        compressed_size, &in_ptr, &out_ptr
+        "Decompress {} bytes from {:?} to {:?}, reserved {:?} bytes\n",
+        compressed_size, &in_ptr, &out_ptr, payload_size
     );
 
     let input = unsafe { slice::from_raw_parts(in_ptr, compressed_size as usize) };
-    let output = unsafe { slice::from_raw_parts_mut(out_ptr, LINUXBOOT_SIZE) };
+    let output = unsafe { slice::from_raw_parts_mut(out_ptr, payload_size) };
 
     let result = MyLzss::decompress(
         lzss::SliceReader::new(input),
         lzss::SliceWriter::new(output),
     );
     match result {
-        Ok(r) => {
-            print!("Success, decompressed {r} bytes :)\n");
-            let a = LINUXBOOT_ADDR + 0x30;
-            let r = unsafe { read_volatile(a as *mut u32) };
-            if r == u32::from_le_bytes(*b"RISC") {
-                print!("Payload looks like Linux Image, yay!\n");
-            } else {
-                panic!("Payload does not look like Linux Image: {:x}\n", r);
-            }
-        }
+        Ok(r) => print!("Success, decompressed {r} bytes :)\n"),
         Err(e) => panic!("Decompression error {e}\n"),
     }
+}
 
+fn decompress_lb() {
```

Comment:
TODO: should this be moving to a library?

Reply:
yea, or rather the other helper first, which is already in shape for that

Reply:
did that - PTAL :)